### ISSUE TITLE
DEBUG=1 compilation fails with nccl 2.16.2

### ISSUE THREAD

**Issue by visweshfb** [NVIDIA/nccl#775]
Compiling nccl with with DEBUG=1 (make DEBUG=1), fails as below. This is with nccl 2.16.2.

Please suggest a work around, if this is a known issue.

make -C src build BUILDDIR=/root/trial_nccl_2.16/build
make[1]: Entering directory '/root/trial_nccl_2.16/src'
NVCC_GENCODE is -gencode=arch=compute_50,code=sm_50 -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_35,code=sm_35 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_80,code=compute_80
make[2]: Entering directory '/root/trial_nccl_2.16/src/collectives/device'
NVCC_GENCODE is -gencode=arch=compute_50,code=sm_50 -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_35,code=sm_35 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_80,code=compute_80
make[2]: Leaving directory '/root/trial_nccl_2.16/src/collectives/device'
Linking    libnccl.so.2.16.2                   > /root/trial_nccl_2.16/build/lib/libnccl.so.2.16.2
/usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crti.o: In function `_init':
(.init+0xb): relocation truncated to fit: R_X86_64_REX_GOTPCRELX against undefined symbol `__gmon_start__'
/bin/ld: failed to convert GOTPCREL relocation; relink with --no-relax
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:80: /root/trial_nccl_2.16/build/lib/libnccl.so.2.16.2] Error 1

I could try to build only for the native arch, if it helps to reduce the foot print.

I am looking to enable cuda-gdb for kernel part of nccl, hence trying to compile with DEBUG-1.

If you could suggest other ways to enable cuda-gdb, that will be useful.

Thanks.

**Comment by jbachan** [comment]
This looks like a platform issue. Some searching and I found this: https://bugzilla.redhat.com/show_bug.cgi?id=1713891

**Comment by AddyLaddy** [comment]
To reduce the binary size you can start by limiting the GENCODE to just the arch you're working on by setting `NVCC_GENCODE`

I also have had some success by just recompiling the kernel module I'm interested in with DEBUG=1. e.g. compile NCCL without DEBUG=1 then do something like `touch src/collectives/device/all_reduce.cu` followed by `make -j DEBUG=1`

**Comment by visweshfb** [comment]
Thanks for the suggestions. I was thinking in the same lines, hence had checked about this in the ticket - 

>I could try to build only for the native arch, if it helps to reduce the foot print.

Is there a way to know the arch for Ampere arch GPUs. i am not sure whether it is compute_80 or not. Also, there seems to be an option to give "native" option, by which the compiler would check the GPU arch in the host, where i am compiling. If it can do that, that would be ideal for me as well.

Thanks for the bugzilla pointer, let me try to compile in a different platform, to see if it makes a difference. 

Regarding - 
>I also have had some success by just recompiling the kernel module I'm interested in with DEBUG=1. e.g. compile NCCL without DEBUG=1 then do something like touch src/collectives/device/all_reduce.cu followed by make -j DEBUG=1

I tried doing something similar, but faced with the same issue as before...

**Comment by sjeaugey** [comment]
> Is there a way to know the arch for Ampere arch GPUs. i am not sure whether it is compute_80 or not. 

Yes it is 80. In doubt, check https://developer.nvidia.com/cuda-gpus.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
